### DISCUSSION TITLE
TypedArray (e.g. Uint8Array) into BufRead/Read without memory copying

### DISCUSSION THREAD

**Discussion by ingun37** [wasm-bindgen/wasm-bindgen#2780] [Q&A] [answered]
I need Uint8Array in the form of [BufRead](https://doc.rust-lang.org/nightly/std/io/trait.BufRead.html) or [Read](https://doc.rust-lang.org/stable/std/io/trait.Read.html).

I could just copy the whole memory using something like [to_vec](https://docs.rs/js-sys/latest/js_sys/struct.Uint8Array.html#method.to_vec) and construct some type of `BufRead`/`Read` but I don't want to because the Uint8Array I'm dealing with can be huge (50~100mb)

I think in order to do that I have to get the address to the mutable continuous memory block, probably in type of `*mut [u8]`. (I'm not familiar with Rust yet so the syntax might not be right but hope you get the idea).
But I can't find the suitable method from [Uint8Array doc](https://docs.rs/js-sys/latest/js_sys/struct.Uint8Array.html).

Is it possible? If so, how can I do it?

**Comment by alexcrichton** [comment, marked as answer]
You'll probably want to hold a `js_sys::Uint8Array` and then develop your own abstraction around it (using methods in `js-sys` to copy data in/out). Currently there's nothing built-in for this, however.
